src/extra/utils/ftd.pyx

```diff
@@ -0,0 +1,517 @@
+    cdef data_t orig_walk = walk
+    if negative:
+        walk = -walk
+
+    # A rounded-up delay is required to control various loops, and also
+    # used instead of the floating delay if its value is actually an
+    # integer. Omitting sinc interpolation in these cases is a huge
+    # performance increase.
+    cdef int int_delay = <int>ceil(delay)
+    cdef bint is_integer_delay = int_delay == delay
+
+    with nogil:
+        for i in range(int_delay, signal.shape[0] - 1):
+            if negative:
+                if signal[i] >= threshold:
+                    continue
+            elif signal[i] <= threshold:
+                continue
+
+            j = i + 1
+
+            if not is_integer_delay:
+                if negative:
+                    cfd_i = fraction * signal[i] - _sinc_interp(
+                        signal, <data_t>i - delay)
+                    cfd_j = fraction * signal[j] - _sinc_interp(
+                        signal, <data_t>j - delay)
+                else:
+                    cfd_i = _sinc_interp(
+                        signal, <data_t>i - delay) - fraction * signal[i]
+                    cfd_j = _sinc_interp(
+                        signal, <data_t>j - delay) - fraction * signal[j]
+            elif negative:
+                cfd_i = fraction * signal[i] - signal[i - int_delay]
+                cfd_j = fraction * signal[j] - signal[j - int_delay]
+            else:
+                cfd_i = signal[i - int_delay] - fraction * signal[i]
+                cfd_j = signal[j - int_delay] - fraction * signal[j]
+
+            # From this point on, the CFD values are computed in such a
+            # way as if the signal and threshold would be positive.
+
+            if cfd_i < walk and cfd_j > walk and i > next_edge:
+                if interp == EdgeInterpolation.NEAREST:
+                    edge_pos = i if fabs(cfd_j - walk) > fabs(cfd_i - walk) \
+                        else j
+                elif interp == EdgeInterpolation.LINEAR:
+                    edge_pos = i + (cfd_i - walk) / (cfd_i - cfd_j)
+                elif interp == EdgeInterpolation.SPLINE:
+                    raise NotImplementedError('spline interpolation')
+                elif interp == EdgeInterpolation.SINC:
+                    # As sinc interpolation still makes use of the
+                    # original (possibly negative!) signal, it also
+                    # requires the original walk value for comparions.
+                    edge_pos = _cfd_sinc_interp(signal, i, j, delay, fraction,
+                                                orig_walk)
+                else:
+                    raise ValueError('invalid interpolation mode')
+
+                edges[edge_idx] = edge_pos
+                next_edge = i + width
+                edge_idx += 1
+
+                if edge_idx == max_edge:
+                    break
+
+    return np.asarray(edges)[:edge_idx], np.asarray(amplitudes)[:edge_idx], \
```

Comment:
Is it needed to convert the return values to numpy arrays? They have to be array-like anyway, so presumably `[:n]` works.

Reply:
Both `edges` as well as `edges[:edge_idx]` yield a `MemoryView` object on the Python side. Cython recommends to use the typed memoryview interface for new code rather than the prior numpy integration, which has the unfortunate side effect you need to cast it back to a Python type at the end. On the upside though, any type implementing the buffer protocol is supported as input this way.

Reply:
Ah, and typed memoryview do not require the GIL when passed around 🙂

Reply:
Does [numpy.asanyarray()](https://numpy.org/doc/stable/reference/generated/numpy.asanyarray.html) work? That should pass numpy-like arrays through unchanged - but I don't know if it also passes MemoryView through.

Reply:
But when should it ever be an ndarray in that place? On the Cython side, it must always be a memory view of something you cannot determine on the Cython side.

Reply:
Right, my mistake. I was thinking in Python terms, where if you pass an array-like object in that's what the function works with. But with the type specification in Cython, it's converted to a memory view whatever you pass in.